### PR TITLE
API: Raise warning when overriding a CRS via crs property setter

### PR DESCRIPTION
**Pull request by JohnMoutafis**
Refs: #3085 
For the current version: Warn the user that tries to manually change a GeoSeries' **existing** CRS by the property setter (ex. `gs.crs = "new crs"`), that the setter will be deprecated in future versions and that the `set_crs` method should be used instead.

### CODE REVIEW

**Comment by martinfleis** [comment]
As noted in https://github.com/geopandas/geopandas/issues/3085, we'll need to cover both GeoSeries and GeoDataFrame here.

**Comment by JohnMoutafis** [comment]
There seems to be an issue with `GeoDataFrame.dissolve` trying to set `crs` through the setter but I cannot wrap my head around were this is happening.
Any suggestions?

File: geopandas/geodataframe.py
Comment (martinfleis) [inline geopandas/geodataframe.py:522]:
```suggestion
            warnings.warn(
                f"Overriding the CRS of a GeoDataFrame that "
                f"already has CRS. "
                f"This unsafe behavior will be deprecated in future versions."
                f"Use GeoDataFrame.set_crs method instead.",
                stacklevel=2,
                category=DeprecationWarning,
```

1. CRS printing can be veery noisy
2. no need to shout :)
3. proper stacklevel

please replicate the same in GeoSeries.

File: geopandas/tests/test_crs.py
Comment (martinfleis) [inline geopandas/tests/test_crs.py:340]: Why are there these changes?

File: geopandas/tests/test_geodataframe.py
Comment (martinfleis) [inline geopandas/tests/test_geodataframe.py:375]: Use `with pytest.warns` instead. Check for other cases of that in tests to see how we use it.

File: geopandas/tests/test_geoseries.py
Comment (martinfleis) [inline geopandas/tests/test_geoseries.py:83]: I understand this is just some form of linting but if you can limit the changes in the PR to relevant lines only, it would help the review.

File: geopandas/tests/test_geoseries.py
Comment (martinfleis) [inline geopandas/tests/test_geoseries.py:527]: Same as above.

**Review by martinfleis — COMMENTED** [review]
I've fixed the usage of GeoSeries.crs setter in GeoSeries.set_geometry and pushed one minor commit to ensure_geometry along the same lines.

A consequence of this warning is that due to usage of setter in some places and especially in tests, we have a ton of there warnings in our CI which we should ideally clear, especially if this will raise eventually.

**Comment by jorisvandenbossche** [comment]
@JohnMoutafis would you have time to update for the comments of @martinfleis ?

**Comment by JohnMoutafis** [comment]
@jorisvandenbossche I will try to find some (hopefully soon)

**Comment by martinfleis** [comment]
@JohnMoutafis given we'd like to have this included in 1.0 which shall be released hopefully soon, I have pushed some commits myself, mostly resolving my own comments.

**Review by martinfleis — APPROVED** [review]
Thanks for getting this going @JohnMoutafis!

**Comment by JohnMoutafis** [comment]
@martinfleis sorry for not finding the time to handle the requested changes.
Does the PR need anything atm to be merged?
What about the coverage?

**Comment by martinfleis** [comment]
I think it is ready to be merged, just want to wait for another review.

The coverage can be ignored I suppose, given the patch has 100%. It tends to be flaky from time to time.

**Review by jorisvandenbossche — APPROVED** [review]
Looks good, thanks!

File: geopandas/geoseries.py
Comment (m-richards) [inline geopandas/geoseries.py:244]: Sorry a bit late now, but these lines are missing spaces after the full stops inside the quotes (this also happens once above in the geodataframe warning). 
I also see in the dev env that the warnings are getting triggered by pandas code in certain tests, which we may want to look into, but I suppose there's still quite a bit of time before the deprecation is enforced.

**Comment by martinfleis** [comment]
Yeah, I noted that in https://github.com/geopandas/geopandas/pull/3088#pullrequestreview-2009085413 and then forgot about it.

> Do we want to accept None in that method, or do we not raise the warning for the case of assigning None ?

I think that we should have a canonical way of removing crs and `set_crs(None, allow_override=True)` feels like the best way.

In the tests, we can potentially replace that by assigning to the `GeometryArray.crs` but allowing None in `set_crs` might be the best.

**Pull request merged**
